### PR TITLE
fix: Persist workflow definition of pending commands on start

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

When the agent restarts and loads pending operations, these operations might not have been assign a workflow version. This is notably the case after a self update from a previous version of the agent that was oblivious of workflow versions.

In such a case, the agent assumes the current user-defined version of the workflow is the version to be used for the pending command. The workflow definition is persisted in `workflows-in-use` directory and assigned to the command.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/3305

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 1]
```diff
@@ -346,7 +346,25 @@ impl WorkflowRepository {
...
+        // If the resumed commands have been triggered by an agent without workflow version management
+        // then these commands are assigned the current version of the operation workflow.
+        // These currents versions have also to be marked as in use and persisted.
+        for (_, ref mut command) in commands.iter_mut() {
+            if command.workflow_version().is_none() {
+                if let Some(operation) = command.operation() {
+                    if let Some(current_version) = self.workflows.use_current_version(&operation) {
+                        self.persist_workflow_definition(&operation, &current_version)
+                            .await;
+                        *command = command.clone().set_workflow_version(&current_version);
```
Comment (jarhodes314) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:362]: I'm puzzled by this API where we seemingly are cloning `command` just to mutate it and return it from `set_workflow_version`?

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 2]
```diff
@@ -150,6 +132,16 @@ impl WorkflowSupervisor {
...
     }
 
+    /// Mark the current version of an operation workflow as being in use.
+    ///
+    /// Return the current version if any.
+    pub fn use_current_version(&mut self, operation: &OperationName) -> Option<WorkflowVersion> {
+        match self.workflows.get_mut(&operation.as_str().into()) {
+            Some(versions) => versions.use_current_version().cloned(),
+            None => None,
+        }
```
Comment (jarhodes314) [inline crates/core/tedge_api/src/workflow/supervisor.rs:142]:
```suggestion
       self.workflows.get_mut(&operation.as_str().into())?.use_current_version().cloned()
```

**Review by jarhodes314 — APPROVED** [review]
I'm a bit puzzled by some of what's going on with the APIs, but believe this functions correctly

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs:362]: You can be!

Not only `set_workflow_version` but all the updating methods of `GenericCommandState` are consuming `self`. 
Fixing that would not be complicated, but I prefer to fix this bug asap.

File: crates/core/tedge_api/src/workflow/supervisor.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs:142]: Good point. I will fixed that in a follow-up PR.

**Pull request merged**